### ISSUE TITLE
Graph-based matching for optimal (and correct) reference to estimate event matching

### ISSUE THREAD

**Issue by justinsalamon** [TUT-ARG/sed_eval#3]
Just found out about this library, great work! (was about to start writing basically the same but decided to have one final look to see if there's anything out there already - glad I found it).

For the event-based metrics the [evaluate function](https://github.com/TUT-ARG/sed_eval/blob/master/sed_eval/sound_event.py#L977) matches events by iterating over all reference and estimate events in a [nested loop](https://github.com/TUT-ARG/sed_eval/blob/master/sed_eval/sound_event.py#L1007). I have two questions about this:
1. Doesn't this mean that the same estimated event can be matched against more than one reference event? I'm not sure that's a desired behavior - presumably every reference event should be matched again at most one estimated event and vice versa?
2. Assuming the goal of (1) is indeed to match every reference event against at most one estimated event (and vice versa), then using a nested loop means the matching is greedy and not necessarily optimal (i.e. it might not find the optimal pairing of matching events) leading to an underestimation of performance. This is exactly the case for note transcription (which is basically the same problem: matching pairs of events based on onset, offset and label (pitch) criteria). To get around this in `mir_eval` we [match notes using bi-partite graph matching](https://github.com/craffel/mir_eval/blob/master/mir_eval/transcription.py#L335), which guarantees to find the optimal pairing of reference and estimated events. The same approach is also used everywhere else in `mir_eval` where to sets of events needs to be matched into pairs.

Let me know what you think, cheers!

**Comment by toni-heittola** [comment]
Thanks for bringing these issues up. 

Yes, you are correct with first issue. The intention in this nested loop is to find correctly detected events, however, only one correctly detected event per reference event is accepted. The rest will be counted as substitutions. 

On the other hand, the code did not implement this idea fully. I fixed the code now (commits 857a0adcfcaec5174e84cc662abfb6f82a81a398 and 686d1603d5fe0ce022572d3a02ce15203ae0526a) so that  estimated event marked already matched cannot be matched again. On rather small test dataset (two files from DCASE2013), this fix did not change much ER metric numbers.

As for the second issue, with current metrics I do not see the need to find best matched event pairs. Since we are only counting whether an event is found or not, if multiple events match within the required conditions, one will be correct and the others counted as insertions (or substitutions), no matter which is best match. In music the pitch tolerance is something you also care for, but in our case that is fixed, and only temporal position matters. We may consider optimal matching later, if we move to some temporal overlap related metrics.

**Comment by justinsalamon** [comment]
> On the other hand, the code did not implement this idea fully. I fixed the code now (commits 857a0ad and 686d160) so that estimated event marked already matched cannot be matched again.

👍

> On rather small test dataset (two files from DCASE2013), this fix did not change much ER metric numbers.

But on a larger set it might, right?

> As for the second issue, with current metrics I do not see the need to find best matched event pairs. Since we are only counting whether an event is found or not, if multiple events match within the required conditions, one will be correct and the others counted as insertions (or substitutions), no matter which is best match. 

Consider the following case (somewhat contrived, but possible):

Reference = 
1, 3, carhorn
1.5, 3, carhorn

Estimate = 
1.25, 3, carhorn
1.75, 3, carhorn

The first estimated event matches both reference events (assuming a tolerance of .25), however the second estimated event only matches the second reference event. With greedy matching you might (if you're unlucky) match the first estimated event to the second reference event, and then the other pair (the second estimated event vs the first reference event) won't match. With graph matching you're guaranteed the first estimated event will be paired with the first reference event and the second with the second.

I doubt the metrics would change dramatically, but it's worth considering.

**Comment by bmcfee** [comment]
> As for the second issue, with current metrics I do not see the need to find best matched event pairs. Since we are only counting whether an event is found or not, if multiple events match within the required conditions, one will be correct and the others counted as insertions (or substitutions), no matter which is best match.

Sorry to jump in here a year and change late, but maybe I can provide some context.  We discuss this broadly in section 4.1 of the [mir_eval paper](http://colinraffel.com/publications/ismir2014mir_eval.pdf).

The reason to care about finding an optimal matching -- note "an" not "the", as they need not be unique -- is for reproducibility.  Greedy matching will generally be sensitive to the order in which observations are presented, and the same set of observations presented in different orders could produce different scores.

In @justinsalamon's example above, if the estimates were presented in reverse order, it would match `(1.75, 1.5)` first, leaving `(1.25, 1)` as a second match (2 hits).  In the order given, it could match `(1.25, 1)` leaving `(1.75, 1)` which would not be a match.  Note that this is phrased here in terms of time "tolerance", not pitch tolerance.  When we implemented this for onset detection and beat tracking, we observed a small, but consistent increase in the resulting scores, indicating that this phenomenon does occur in real data, and previous implementations were under-reporting accuracy metrics.

**Comment by toni-heittola** [comment]
After a long time, I managed to find time to address this issue. Graph based event matching is added with commits 13cef668f55f0ea7a47cb6b8c719ac86db75b708 and f4726a603601758413f3b0317c94711036656742, and is available in the release version 0.2.1. 

Previously used greedy matching approach is still available though optional parameter `event_matching_type='greedy'` for backward compatibility. The greedy matching approach was implemented such a way that it was sensitive to the order of events in the reference event list. This could produce slightly different metric scores, if event order is changed in cases with events having onsets very close each other (distance smaller than `t_collar`) and assigned to same event label.
